rest_framework/renderers.py

```diff
@@ -342,8 +342,24 @@ def render_field(self, field, parent_style):
         # Get a clone of the field with text-only value representation.
         field = field.as_form_field()
 
-        if style.get('input_type') == 'datetime-local' and isinstance(field.value, str):
-            field.value = field.value.rstrip('Z')
+        if style.get('input_type') == 'datetime-local':
+            has_format = hasattr(field._field, "format")
+
+            if has_format:
+                if field._field.format is None:
+                    datetime_field_value = field.value
+                else:
+                    datetime_field_value = datetime.datetime.strptime(field.value, field._field.format)
+            else:
+                if api_settings.DATETIME_FORMAT == ISO_8601:
+                    datetime_field_value = datetime.datetime.fromisoformat(field.value.rstrip('Z'))
+                else:
+                    datetime_field_value = datetime.datetime.strptime(field.value, api_settings.DATETIME_FORMAT)
+
+            # The format of an input type="datetime-local" is "yyyy-MM-ddThh:mm"
+            # followed by optional ":ss" or ":ss.SSS", so keep only the first three
+            # digits of milliseconds to avoid browser console error.
+            field.value = datetime_field_value.replace(tzinfo=None).isoformat(timespec="milliseconds")
```

Comment:
If `format=None`, then `datetime_field_value` gets set to `field.value`, which apparently may be a string. (In fact, the previous code checked for that in line 345.)

However, this line assumes that `datetime_field_value` is a datetime object.

I'm not sure under which circumstances the expression would be a string. The code is only correct if `field.value` never is a string when `format=None`. Is that guaranteed? (How?) If not, there is a bug here.

The type check was introduced in https://github.com/encode/django-rest-framework/commit/df7b6fcf58417fd95e49655eb140b387899b1ceb#diff-d969c1c82f7a9d0c9b55cd44872d2ad719203180f1404e974e6daaf33be020ebR387.